compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.m

```diff
@@ -52,176 +25,101 @@ @interface UIViewController(CMPUIKitUtilsPrivate)
-    [super removeFromParentViewController];
-
-    [self cmp_setNeedsVerifyHierarchy];
 }
 
-- (void)viewDidLoad {
-    if ([self.view isKindOfClass:[CMPView class]]) {
-        __weak typeof(self) weakSelf = self;
-        ((CMPView *)self.view).windowDidChangeBlock = ^{
-            [weakSelf cmp_setNeedsVerifyHierarchy];
-        };
-    } else {
-        NSAssert(false, @"The view of CMPViewController should be either CMPView or subclassed from CMPView");
+/// This method can be triggered by both `viewWillAppear` and `didMoveToParentViewController`. Mind that in second case the `view` could still not be constructed (adding `UIViewController`  to inactive tab of `UITabBarController`, for example), so the `ComposeScene` is not neccessarily created at this point.
+- (void)transitLifecycleToStarted {
+    switch (_lifecycleState) {
+        case CMPViewControllerLifecycleStateDestroyed:
+            @throw [NSException exceptionWithName:@"CMPViewControllerMisuse"
+                                           reason:@"CMPViewController shouldn't be reused after completely removed from hierarchy, because it's logically marked as Destroyed. You must create a new CMPViewController and use it instead."
+                                         userInfo:nil];
+        case CMPViewControllerLifecycleStateInitalized:
+            _lifecycleState = CMPViewControllerLifecycleStateStarted;
+            [self viewControllerDidEnterWindowHierarchy];
+            [self scheduleHierarchyContainmentCheck];
+            break;
+        case CMPViewControllerLifecycleStateStarted:
+            break;
     }
-
-    [super viewDidLoad];
-}
-
-- (void)loadView {
-    self.view = [CMPView new];
 }
 
-- (void)cmp_setNeedsVerifyHierarchy {
-    if (self.cmp_needsVerifyHierarchy) {
-        return;
-    }
-    self.cmp_needsVerifyHierarchy = true;
-
-    __weak typeof(self) weakSelf = self;
-    dispatch_async(dispatch_get_main_queue(), ^{
-        weakSelf.cmp_needsVerifyHierarchy = false;
-        [weakSelf cmp_verifyHierarchyAndScheduleNotificationIfNeeded];
+- (void)scheduleHierarchyContainmentCheck {
+    double delayInSeconds = 0.5;
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{        
+        switch (self->_lifecycleState) {
+            case CMPViewControllerLifecycleStateInitalized:
+            case CMPViewControllerLifecycleStateDestroyed:
+                assert(false);
+                break;
+            case CMPViewControllerLifecycleStateStarted:
+                // perform check
+                if ([self cmp_isInWindowHierarchy]) {
+                    // everything is fine, schedule next one
+                    [self scheduleHierarchyContainmentCheck];
+                } else {
+                    self->_lifecycleState = CMPViewControllerLifecycleStateDestroyed;
+                    [self viewControllerDidLeaveWindowHierarchy];
```

Comment:
I'd wait for at least one issue from the user with `CMPViewControllerMisuse` exception to see the specific scenario. If a user detaches the view controller to reattach it asynchronously, this one async check would not suffice, because we can't really know whether the async check block will be executed before the async block where it's reattached or after, so:
1. A user is using the API not in a way it's supposed to be used.
2. It still doesn't fix the problem caused by this misuse correctly and exhaustively.